packages/webpack-loader/src/compiled-loader.tsx

```diff
@@ -0,0 +1,125 @@
+import path from 'path';
+import { transformFromAstAsync, parseAsync } from '@babel/core';
+import { getOptions } from 'loader-utils';
+import type { CompiledLoaderOptions, LoaderThis } from './types';
+import { CompiledExtractPlugin } from './extract-plugin';
+import { predicate } from './utils/predicate';
+
+/**
+ * Returns user configuration.
+ *
+ * @param loader
+ * @returns
+ */
+function getLoaderOptions(context: LoaderThis<CompiledLoaderOptions>) {
+  const options: CompiledLoaderOptions =
+    typeof context.getOptions === 'undefined'
+      ? // Webpack v4 flow
+        getOptions(context)
+      : // Webpack v5 flow
+        context.getOptions({
+          type: 'object',
+          properties: {
+            importReact: {
+              type: 'boolean',
+            },
+            nonce: {
+              type: 'string',
+            },
+            extract: {
+              type: 'boolean',
+            },
+          },
+        });
+
+  return options;
+}
+
+/**
+ * Compiled webpack loader.
+ *
+ * @param this
+ * @param code
+ */
+export default async function compiledLoader(
+  this: LoaderThis<CompiledLoaderOptions>,
+  code: string
+): Promise<void> {
+  const callback = this.async();
```

Comment:
the webpack loader-runner accepts promises from loader functions:

https://github.com/webpack/loader-runner/blob/6221befd031563e130f59d171e732950ee4402c6/lib/LoaderRunner.js#L138

so you can remove `this.async()` and just return a string e.g.

```js
export default async function doSth() {
  const result = await somethingSlow();
  return result;
}
```

Reply:
oh lovely, thanks

Reply:
how does that play with source maps?